modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx

```diff
@@ -0,0 +1,485 @@
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import ClayButton from '@clayui/button';
+import ClayDropDown from '@clayui/drop-down';
+import ClayForm, {ClayInput} from '@clayui/form';
+import ClayIcon from '@clayui/icon';
+import ClayLayout from '@clayui/layout';
+import ClayModal from '@clayui/modal';
+import {FrontendDataSet} from '@liferay/frontend-data-set-web';
+import classNames from 'classnames';
+import {fetch, openModal} from 'frontend-js-web';
+import fuzzy from 'fuzzy';
+import React, {useEffect, useRef, useState} from 'react';
+
+import '../css/FDSEntries.scss';
+
+const FUZZY_OPTIONS = {
+	post: '</strong>',
+	pre: '<strong>',
+};
+
+type HeadlessResource = {
+	bundleLabel: string;
+	entityClassName: string;
+	name: string;
+	version: string;
+};
+
+interface IHeadlessResourceItemProps {
+	headlessResource: HeadlessResource;
+	query: string;
+}
+
+const HeadlessResourceItem = ({
+	headlessResource,
+	query,
+}: IHeadlessResourceItemProps) => {
+	const fuzzyNameMatch = fuzzy.match(
+		query,
+		headlessResource.name,
+		FUZZY_OPTIONS
+	);
+
+	const fuzzyBundleLabelMatch = fuzzy.match(
+		query,
+		headlessResource.bundleLabel,
+		FUZZY_OPTIONS
+	);
+
+	return (
+		<ClayLayout.ContentRow className="headless-resource">
+			{fuzzyNameMatch ? (
+				<span
+					dangerouslySetInnerHTML={{
+						__html: fuzzyNameMatch.rendered,
+					}}
+				/>
+			) : (
+				<span>{headlessResource.name}</span>
+			)}
+
+			<span className="context">
+				{fuzzyBundleLabelMatch ? (
+					<span
+						dangerouslySetInnerHTML={{
+							__html: fuzzyBundleLabelMatch.rendered,
+						}}
+					/>
+				) : (
+					<span>{headlessResource.bundleLabel}</span>
+				)}
+
+				{` ${headlessResource.version}`}
+			</span>
+		</ClayLayout.ContentRow>
+	);
+};
+
+const RequiredMark = () => (
+	<>
+		<span className="inline-item-after reference-mark text-warning">
+			<ClayIcon symbol="asterisk" />
+		</span>
+		<span className="hide-accessible sr-only">
+			{Liferay.Language.get('required')}
+		</span>
+	</>
+);
+
+interface IDropdownMenuProps {
+	headlessResources: Array<HeadlessResource>;
+	setHeadlessResourceValidationError: Function;
+	setSelectedHeadlessResource: Function;
+}
+
+const DropdownMenu = ({
+	headlessResources: initialHeadlessResources,
+	setHeadlessResourceValidationError,
+	setSelectedHeadlessResource,
+}: IDropdownMenuProps) => {
+	const [headlessResources, setHeadlessResources] = useState<
+		Array<HeadlessResource>
+	>(initialHeadlessResources || []);
+	const [query, setQuery] = useState('');
+
+	const onSearch = (query: string) => {
```

Comment:
We load all headless resources at once from the backend. So we won't flood server with search queries on headless resources. Still, we should consider `throttle()` ing this and run it when `query.lenght>1`. Not needed for the PR to be fwd, just keep it in mind

Reply:
I tried implementing debouncing, but there is a complication with TS and our `debounce` util. We [allow](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/index.d.ts#L74) only parameters of type `() => void`, while our search function has the search text argument. There may a workaround, there may be a good reason for this restriction, or I am maybe missing something here. I'd definitely leave this as a separate task. cc @kresimir-coko

Reply:
IIRC we used `() => void` as a catch-all type for functions without arguments. Not sure if there's a workaround that could fit our case. Maybe @bryceosterhaus has something in mind, seeing as you worked more with TS than me or Marko.

Reply:
Yeah, the problem is that our `onSearch` function, that we want to debounce, is not without arguments.